alexandria/core/migrations/0012_tag_uuid_schema.py

```diff
@@ -0,0 +1,91 @@
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("alexandria_core", "0011_tag_uuid"),
+    ]
+
+    operations = [
+        # indexes and constraints have to be dropped before changing the data type
+        # drop primary key constraint
+        migrations.RunSQL(
+            "ALTER TABLE alexandria_core_tag DROP CONSTRAINT alexandria_core_tag_pkey CASCADE;"
+        ),
+        # drop slug index on tag table and tag_id index on document-tag table
+        # we do not know the exact name of the index, so we have to query it
+        migrations.RunSQL(
+            """
+        do $$
+        declare
+            r record;
+        begin
+            for r in
+                select
+                    t.relname as table_name,
+                    i.relname as index_name,
+                    a.attname as column_name
+                from
+                    pg_class t,
+                    pg_class i,
+                    pg_index ix,
+                    pg_attribute a
+                where
+                    t.oid = ix.indrelid
+                    and i.oid = ix.indexrelid
+                    and a.attrelid = t.oid
+                    and a.attnum = ANY(ix.indkey)
+                    and t.relkind = 'r'
+                    and i.relname like '%like'
+                    and ((
+                        t.relname = 'alexandria_core_tag' and a.attname = 'slug'
+                    )or (
+                        t.relname = 'alexandria_core_document_tags' and a.attname = 'tag_id'
+                    ))
+
+            loop
+                EXECUTE format('DROP INDEX %I;', r.index_name);
+            end loop;
+        end;
+        $$ language plpgsql;
```

Comment:
Doing this PL/pgSQL command was the only way I found to be able to delete the indexes created by django, as they always have a random string in the index name.

Reply:
I honestly dislike the SQL approach, even though I understand it's probably the fastest (or even only) way. It does run the risk of having the DB run out of sync with what Django thinks is there, primarily the indexes.

Tried to produce a django-only version, sadly I was unable to get it to work.

Just to make sure we don't miss any constraints / indexes, Could you do the following?

* Dump the DB schema after this migration (only schema, no data)
* Remove all migrations and create them anew, without any magic
* Drop the database and migrate with the "clean" one
* Diff the DB schemas
* Add any missing indexes / constraints, ideally using django's naming convention, or document that this has been done and the indexes are indeed clean (comment on this PR is enough for me)

I'm afraid if we don't do this, we may run into some consistency issues further down the road that may be very hard to track down / debug

Reply:
After doing this the diff is the foreign key index with the django generated id.

If the entire migration is done in the PL/pgSQL part the index could probably be recreated with the same name.
Other choices would be to create temporary tables where all data is copied over and fresh tables are created, so that django has control over it.

```diff
 --
--- Name: alexandria_core_document_tags alexandria_core_docu_tag_id_fk_alexandri; Type: FK CONSTRAINT; Schema: public; Owner: alexandria
+-- Name: alexandria_core_document_tags alexandria_core_docu_tag_id_8dc7fb00_fk_alexandri; Type: FK CONSTRAINT; Schema: public; Owner: alexandria
 --
 
 ALTER TABLE ONLY public.alexandria_core_document_tags
-    ADD CONSTRAINT alexandria_core_docu_tag_id_fk_alexandri FOREIGN KEY (tag_id) REFERENCES public.alexandria_core_tag(id) DEFERRABLE INITIALLY DEFERRED;
+    ADD CONSTRAINT alexandria_core_docu_tag_id_8dc7fb00_fk_alexandri FOREIGN KEY (tag_id) REFERENCES public.alexandria_core_tag(id) DEFERRABLE INITIALLY DEFERRED;
```

[clean_schema.txt](https://github.com/projectcaluma/alexandria/files/13661861/clean_schema.txt)
[migrated_schema.txt](https://github.com/projectcaluma/alexandria/files/13661862/migrated_schema.txt)

Reply:
Nice! I think it'd be a good idea to name the new constraint exactly the same way as Django would do it, just so Django can find it again in the future (Assuming the suffix is a hash, not purely random).

Reply:
Seems to be deterministic, I think we can try to delete the name with the the fixed name, instead of the pgsql
https://github.com/django/django/blob/main/django/db/backends/utils.py#L302
https://github.com/django/django/blob/8fa7c2ae88aee7a2b54a745d25ed38152aad2591/django/db/models/indexes.py#L183C24-L183C36
Ill change the index deletion

Reply:
[migrated_new_schema.txt](https://github.com/projectcaluma/alexandria/files/13662759/migrated_new_schema.txt)
Now there is no more difference to the django created initial schema.